libr/core/cmd_api.c

```diff
@@ -185,114 +186,251 @@ R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
+#endif
+
+	return ht_pp_update (cmd->aliases, strdup (k), val);
 }
 
-R_API bool r_cmd_alias_del (RCmd *cmd, const char *k) {
-	int i; // find
-	for (i = 0; i < cmd->aliases.count; i++) {
-		if (!k || !strcmp (k, cmd->aliases.keys[i])) {
-			R_FREE (cmd->aliases.values[i]);
-			cmd->aliases.count--;
-			if (cmd->aliases.count > 0) {
-				if (i > 0) {
-					free (cmd->aliases.keys[i]);
-					cmd->aliases.keys[i] = cmd->aliases.keys[0];
-					free (cmd->aliases.values[i]);
-					cmd->aliases.values[i] = cmd->aliases.values[0];
-				}
-				memmove (cmd->aliases.values,
-					cmd->aliases.values + 1,
-					cmd->aliases.count * sizeof (void*));
-				memmove (cmd->aliases.keys,
-					cmd->aliases.keys + 1,
-					cmd->aliases.count * sizeof (void*));
+R_API int r_cmd_alias_set_raw(RCmd *cmd, const char *k, const ut8 *v, int sz) {
+	RCmdAliasVal *val = R_NEW (RCmdAliasVal);
+	val->data = malloc (sz);
+	memcpy (val->data, v, sz);
+	val->sz = sz;
+
+	/* If it's a string already, we speed things up later by checking now */
+	const ut8 *firstnull = NULL;
+	for (int i = 0; i < sz; i++) {
+		/* \0 before expected -> not string */
+		if (v[i] == '\0') {
+			firstnull = &v[i];
+			break;
+		}
+	}
+
+	if (firstnull && firstnull == &v[sz-1]) {
+		/* Data is already a string */
+		val->is_str = true;
+#if DEBUG_ALIAS
+		eprintf ("Alias data is already a string.\n");
+#endif
+	} else if (!firstnull) {
+		/* Data is an unterminated string */
+		val->sz++;
+		val->data = realloc(val->data, val->sz);
+		val->data[val->sz-1] = '\0';
+		val->is_str = true;
+#if DEBUG_ALIAS
+		eprintf ("Alias data was unterminated string, added \\0.\n");
+#endif
+	} else {
+		val->is_str = false;
+#if DEBUG_ALIAS
+		eprintf ("Alias data is bytes.\n");
+#endif
+	}
+
+	val->is_data = true;
+
+	if (val->is_str) {
+		/* No trailing newline */
+		int len = val->sz - 1;
+		while (len-- > 0) {
+			if (v[len] == '\r' || v[len] == '\n') {
+				val->data[len] = '\0';
+			} else {
+				break;
 			}
-			return true;
 		}
+		val->sz = len+1;
 	}
-	return false;
+
+#if DEBUG_ALIAS
+	if (val->is_str) {
+		char *v_str = r_cmd_alias_val_strdup (val);
+		eprintf ("Creating raw (string) alias (encoded): "
+				"\"%s\" -> \"%s\"\n", k, v_str);
+		free (v_str);
+	} else {
+		char *escaped = r_cmd_alias_val_strdup (val);
+		eprintf ("Creating new raw alias (escaped): "
+				"\"%s\" -> \"%s\"\n", k, escaped);
+		free (escaped);
+	}
+#endif
+
+	return ht_pp_update (cmd->aliases, strdup (k), val);
 }
 
-// XXX: use a hashtable or any other standard data structure
-R_API int r_cmd_alias_set(RCmd *cmd, const char *k, const char *v, int remote) {
-	void *tofree = NULL;
-	if (!strncmp (v, "base64:", 7)) {
-		ut8 *s = r_base64_decode_dyn (v + 7, -1);
-		if (s) {
-			tofree = s;
-			v = (const char *)s;
-		}
+R_API RCmdAliasVal *r_cmd_alias_get(RCmd *cmd, const char *k) {
+	r_return_val_if_fail (cmd && cmd->aliases && k, NULL);
+
+	bool found;
+	RCmdAliasVal *v = ht_pp_find(cmd->aliases, k, &found);
+#if DEBUG_ALIAS
```

Comment:
Maybe we can get rid of all this debugging code?

Reply:
That's fine with me